packages/dapi/lib/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.js

```diff
@@ -0,0 +1,189 @@
+ * @param {CoreRpcClient} coreAPI
+ * @param {ZmqClient} zmqClient
+ * @param {subscribeToNewBlockHeaders} subscribeToNewBlockHeaders
+ * @return {subscribeToBlockHeadersWithChainLocksHandler}
+ */
+function subscribeToBlockHeadersWithChainLocksHandlerFactory(
+  getHistoricalBlockHeadersIterator,
+  coreAPI,
+  zmqClient,
+  subscribeToNewBlockHeaders,
+) {
+  /**
+   * @typedef subscribeToBlockHeadersWithChainLocksHandler
+   * @param {grpc.ServerWriteableStream<BlockHeadersWithChainLocksRequest>} call
+   */
+  async function subscribeToBlockHeadersWithChainLocksHandler(call) {
+    const { request } = call;
+
+    let fromBlockHash = Buffer.from(request.getFromBlockHash()).toString('hex');
+    const fromBlockHeight = request.getFromBlockHeight();
+    const count = request.getCount();
+
+    const newHeadersRequested = count === 0;
+
+    const acknowledgingCall = new AcknowledgingWritable(call);
+
+    const mediator = new ProcessMediator();
+
+    mediator.on(
+      ProcessMediator.EVENTS.BLOCK_HEADERS,
+      async (blockHeaders) => {
+        await sendBlockHeadersResponse(acknowledgingCall, blockHeaders);
+      },
+    );
+
+    mediator.on(
+      ProcessMediator.EVENTS.CHAIN_LOCK,
+      async (chainLock) => {
+        await sendChainLockResponse(acknowledgingCall, chainLock);
+      },
+    );
+
+    if (newHeadersRequested) {
+      subscribeToNewBlockHeaders(mediator, zmqClient, coreAPI);
+    }
+
+    // If block height is specified instead of block hash, we obtain block hash by block height
+    if (fromBlockHash === '') {
+      if (fromBlockHeight === 0) {
+        throw new InvalidArgumentGrpcError('minimum value for `fromBlockHeight` is 1');
+      }
+
+      // we don't need to check bestBlockHeight because getBlockHash throws
+      // an error in case of wrong height
+      try {
+        fromBlockHash = await coreAPI.getBlockHash(fromBlockHeight);
+      } catch (e) {
+        if (e.code === -8) {
+          // Block height out of range
+          throw new NotFoundGrpcError('fromBlockHeight is bigger than block count');
+        }
+
+        throw e;
+      }
+    }
+
+    let fromBlock;
+
+    try {
+      fromBlock = await coreAPI.getBlock(fromBlockHash);
+    } catch (e) {
+      // Block not found
+      if (e.code === -5) {
+        throw new NotFoundGrpcError('fromBlockHash is not found');
+      }
+
+      throw e;
+    }
+
+    if (fromBlock.confirmations === -1) {
+      throw new NotFoundGrpcError(`block ${fromBlockHash} is not part of the best block chain`);
+    }
+
+    const bestBlockHeight = await coreAPI.getBestBlockHeight();
+
+    const historicalCount = count === 0 ? bestBlockHeight - fromBlock.height + 1 : count;
+
+    if (fromBlock.height + historicalCount > bestBlockHeight + 1) {
+      throw new InvalidArgumentGrpcError('`count` value exceeds the chain tip');
+    }
+
+    const bestChainLock = await coreAPI.getBestChainLock();
```

Comment:
Chainlock might not be present. It's possible if core node is just started. You need to send it in case if you have it only.

Reply:
Done